src/dos/cdrom.cpp

```diff
@@ -261,3 +262,212 @@ void CDROM_Interface::LagDriveResponse() const
 
 	prev_ticks = PIC_Ticks;
 }
+
+void CDROM_Interface_Physical::CdReaderLoop()
+{
+	std::unique_lock lock(mutex);
+	while (true) {
+		waiter.wait(lock, [&]{ return sectors_remaining > 0 || should_exit; });
+
+		if (should_exit) {
+			return;
+		}
+
+		if (!queue.IsRunning()) {
+			queue.Clear();
+			queue.Start();
+		}
+
+		// Copy these to local variables so that we can release the mutex lock.
+		const uint32_t local_sectors_remaining = sectors_remaining;
+		const uint32_t local_sector_start = current_sector;
+		lock.unlock();
+
+		// Slow call to read from the CDROM drive. Avoid holding any locks
+		const std::vector<int16_t> audio_samples = ReadAudio(local_sector_start, local_sectors_remaining);
+		const size_t num_frames = audio_samples.size() / 2;
+		const size_t sectors_read = audio_samples.size() / SAMPLES_PER_REDBOOK_FRAME;
+
+		// Convert from int16_t to float before enqueuing
+		std::vector<AudioFrame> audio_frames;
+		for (size_t i = 0; i < num_frames; ++i) {
+			const int16_t left = host_to_le16(audio_samples[i * 2]);
+			const int16_t right = host_to_le16(audio_samples[(i * 2) + 1]);
+			audio_frames.emplace_back(left, right);
+		}
+
+		queue.BulkEnqueue(audio_frames, audio_frames.size());
+
+		lock.lock();
+		if (mixer_channel && !is_paused && sectors_remaining > 0 && queue.IsRunning()) {
+			mixer_channel->Enable(true);
+		}
+		if (sectors_remaining == local_sectors_remaining && current_sector == local_sector_start) {
```

Comment:
Is `>=` a bit safer perhaps?

Reply:
No, that would be wrong.  This is checking to see if the member variables were changed from the main thread while the worker thread had released the mutex lock during reading/enqueuing (ex. the game wants to switch tracks).

This is there to avoid a race condition.  We don't want to modify this variable unless we know it has not changed.

I had a comment explaining this but I must have removed it while I was fixing stuff up.  I'll add one back.

Reply:
Cool, yeah some explanation would help.

Reply:
Comment added